# initial imports

In [1]:
import numpy as np
import pandas as pd
from plotly import  graph_objects as go

In [8]:
from plotly.offline import init_notebook_mode

init_notebook_mode(connected=True)

<IPython.core.display.Javascript object>

In [9]:
%load_ext nb_black

The nb_black extension is already loaded. To reload it, use:
  %reload_ext nb_black


<IPython.core.display.Javascript object>

In [10]:
#!pip install jupyter_contrib_nbextensions
# Configure the nbextensions and install it for the user (doc available in internet)
#!pip install nb_black

<IPython.core.display.Javascript object>

#  This notebook is in Github!
[Notebook link in Github : https://github.com/SarnathK/gp_tutorial](https://github.com/SarnathK/gp_tutorial)

# Authoritative reference book for Gaussian process

## Hear it from the founders

## The Rasmussen and williams book - RW.pdf

### Free download!

[RW.pdf : Rassmussen & williams book - RW.pdf](http://gaussianprocess.org/gpml/chapters/RW.pdf)

# Univariate discrete Distributions

**Rolling of dice**
```
    P(1) = 1/6
    P(2) = 1/6
    ...
    P(6) = 1/6

Fair dice -> Probability mass function (PMF)!
```

**Bernoulli distribution**
```
    Define event: Rolling of dice results in a prime number (2,3,5)
    P(e) in 1 roll = 1/2
    P(not e) in 1 r roll = 1/2
    
    For bernoulli -> You always start with an event with a known probability P(event)
    
    The bernoulli distribution:
        In N rolls of dice, what is the probability of the event occurence?
        Event occuring once, Event occuring twice, ..... , Even occurring N times
        Define probability for each.
        That is a discrete distribution
```

**Geometric distribution**
```
    Number of times you failed the parking test in Dubai before getting the License from RTA
    Thats a geometric distribution.
    Again you start with P(event of success) say 0.1
    And then calculate probability of success after N failures.
    i.e. 0.9^K * 0.1 = P(success at the K+1th attempt)
    This is a PMF
```

# Univariate gaussian distribution formula

## Continuous random variables have Density functions

'x' is a random variable which is centered around the "mean" value

e.g. x could be Height of students in a class

![Univariate gaussian distribution](./img/gaussian_formula.jpg)

# Multivariate discrete distributions

### (discrete) Joint probability distribution of "Rainy" and "Weekdays"

|<span style="color:red">Rainy</span>| Monday| Tuesday | Wednesday | Thursday | Friday |<span style="color:blue">Total</span>|
|------   |--------|----|     -----     |    ----- | -----  | ---- |
| Yes     | 0.02 | 0.03 | 0.10 | 0.20 | 0.07 | 0.42 |
| No      | 0.18 | 0.17 | 0.10 | 0    | 0.13 | 0.58 |
| <span style="color:blue">Total </span>   | 0.20 | 0.20 | 0.20 | 0.20 | 0.20 | 0.20 |

### Conditioning 


**<span style="color:green">If it rained, what is the probability that it was a Tuesday??</span>**

<br>


|<span style="color:red">Rainy</span>| Monday| Tuesday | Wednesday | Thursday | Friday |<span style="color:blue">Total</span>|
|------   |--------|----|     -----     |    ----- | -----  | ---- |
| Yes     | 0.02 | 0.03 | 0.10 | 0.20 | 0.07 | 0.42 |

<br>

**Calulating probability...** 
<br>
```
= 0.03 / (0.02 + 0.03 + 0.10 + 0.20 + 0.07)
= 0.03 / 0.42
= ~7%
```
<br>

**This is same as bayes formula**
<br>
```
P(Tuesday | Rainy) = P(Tuesday and Rainy) / P(Rainy)
```
<br>

### Marginalize on "Rainy"

|<span style="color:red">Rainy</span>| Marginalized probability |
|------   | ---- |
| Yes     | 0.42 |
| No      | 0.58 |

### Marginalize on "Weekdays"
| Monday| Tuesday | Wednesday | Thursday | Friday |
|--------|----|     -----     |    ----- | -----  |
| 0.2 | 0.2 | 0.2| 0.2 | 0.2 |



# Multivariate gaussian joint distribution


## Analytical formula
### Courtesy: Wikipedia

![Multivariate gaussian joint distribution PDF analytical formula](./img/mvg_formula.jpg)

## Gaussians are closed under Conditioning and Marginalization
#### [Click here for full mathematical proof on this subject from a PhD student](https://fabiandablander.com/statistics/Two-Properties.html)

### Marginalizing a gaussian distribution always returns a gaussian distribution
Let **x** and **y** be 2 gaussian multivariate normal vectors...
![Joint distribution of 2 random vectors X and Y](img/gaussian_joint_x_y.jpg)
Then if you marginalize for **x** then:
![Marginalized x](img/x_marginalized.jpg)
#### How to calculate marginal distributions?
<br>
Integrate out the dimensions that are not needed and keep only the dimensions that are needed
<br>
The easier proof is to see that if "x" is multivariate gaussian vector, and if A is some arbitrary "matrix" then A.x is also a gaussian. One can use this lemma to come up with a A that extracts only the first K components from X and thus the subset of that X will also be a gaussian. 

You can see the same proof in <br>

[reference URL set in the header](https://fabiandablander.com/statistics/Two-Properties.html)

You can also see this from <br>

[youtube video from Math monk]('https://www.youtube.com/watch?v=ycDSJkZ_h0I') 

### Conditional distributions of Multivariate gaussians

```
Conditional distribution (bayes conditional probability)

P(X1,X2,X3,...,Xk | Y1,Y2,Y3,...Ym) 
              = 
P(X1, X2, X3, .......Xk, Y1, Y2, ..... Ym) / P(Y1, Y2, ...., Ym)

The denominator can be calculated from the marginal distribution of Y by substituting "Y" as "Y1, Y2, ... Ym"

The numerator is the joint probability distribution of X and Y where Y is substited for "Y1, Y2, ... Ym"

Thus the term yields the distribution of X1, X2, ..... Xk

This distribution happens to be Gaussian!
```
For a mathematical proof: <br>

[Proof that gaussian is closed under conditioning](https://fabiandablander.com/statistics/Two-Properties.html)

#### Analytical formula here from RW.pdf:
![Conditional distribution formula for gaussians](img/gaussian_conditional_formula.jpg)

# Multivariate distribution : 2D distribution

In [11]:
np.random.multivariate_normal([0, 0], [[1, 0], [0, 1]])

array([-0.19279694,  1.03681602])

<IPython.core.display.Javascript object>

In [12]:
np.random.multivariate_normal([0, 0], [[1, 0], [0, 1]])

array([ 0.79003282, -0.04460392])

<IPython.core.display.Javascript object>

## Random sample 2D: Scatter plot

In [13]:
def make_xs_ys(covm=[[1, 0], [0, 1]]):
    xs = []
    ys = []
    for _ in range(1000):
        x, y = np.random.multivariate_normal([0, 0], covm)
        xs.append(x)
        ys.append(y)
    return xs, ys

<IPython.core.display.Javascript object>

In [14]:
def go_scatter(xs, ys, title):
    fig = go.Figure()

    fig.add_trace(go.Scatter(x=xs, y=ys, mode="markers"))

    fig.update_layout(title=title)
    return fig

<IPython.core.display.Javascript object>

In [15]:
xs, ys = make_xs_ys()

<IPython.core.display.Javascript object>

In [16]:
go_scatter(xs, ys, "2D random sample with no correlation")

<IPython.core.display.Javascript object>

## Correlated 2D random samples

In [24]:
def get_corr_cov():
    # Generate independent random normal numbers...
    somex = np.random.normal(0, 1, 1000)
    somey = np.random.normal(0, 1, 1000)
    # Create correlated variables from the 2 independent random variables
    corx = 2 * somex + 3 * somey
    cory = 5 * somex + 2 * somey
    return np.cov(corx, cory)

<IPython.core.display.Javascript object>

In [25]:
cov_matrix = get_corr_cov()
cov_matrix

array([[13.16320429, 16.0789857 ],
       [16.0789857 , 28.26308702]])

<IPython.core.display.Javascript object>

In [26]:
xs_c, ys_c = make_xs_ys(cov_matrix)

<IPython.core.display.Javascript object>

In [31]:
fig = go_scatter(xs_c, ys_c, "Correlated random 2D samples")
fig

<IPython.core.display.Javascript object>

## (Empirical) Conditioning of a correlated 2D distribution

In [50]:
def make_y_for_x(given_x, mean=[0, 0], cov_matrix=[[1, 0], [0, 1]], tol=0.01):
    n = 0
    xs = []
    ys = []
    while n < 100:
        x, y = np.random.multivariate_normal(mean, cov_matrix)
        if np.abs(x - given_x) <= tol:
            xs.append(x)
            ys.append(y)
            n += 1
    return xs, ys

<IPython.core.display.Javascript object>

In [280]:
cov_matrix = get_corr_cov()

<IPython.core.display.Javascript object>

In [284]:
xs_c_m, ys_c_m = make_y_for_x(3, cov_matrix=cov_matrix)
fig = go_scatter(xs_c_m, ys_c_m, title="Correlated 2D samples at x ~ 3")
fig.update_xaxes(range=[-4, 4])
fig.update_yaxes(range=[-15, 15])

<IPython.core.display.Javascript object>

In [285]:
xs_c_m, ys_c_m = make_y_for_x(-3, cov_matrix=cov_matrix)
fig = go_scatter(xs_c_m, ys_c_m, title="Correlated 2D samples at x ~ -3")
fig.update_xaxes(range=[-4, 4])
fig.update_yaxes(range=[-10, 10])

<IPython.core.display.Javascript object>

# Functions as random samples

In [286]:
def get_nd_cov(d):
    #
    # Return a positive-definite matrix : A^T*A is positive-definite for any A
    # NOTE:
    # Positive definite does not mean each entry is positive....!
    # It means that all its eigen values are real and positive....
    #
    m1 = np.matrix(np.random.normal(0, 20, d * d)).reshape(d, d)
    m2 = np.transpose(m1).dot(m1)
    return m2

<IPython.core.display.Javascript object>

In [289]:
m1 = get_nd_cov(50)
assert (m1 == m1.T).sum() == 50 * 50

<IPython.core.display.Javascript object>

In [287]:
eivals, _ = np.linalg.eig(get_nd_cov(50))
eivals

array([7.18207788e+04, 6.58449247e+04, 6.02644379e+04, 5.74239372e+04,
       5.44743839e+04, 5.12625889e+04, 5.01343072e+04, 4.45228406e+04,
       4.11870318e+04, 4.00127899e+04, 3.83281950e+04, 3.44079834e+04,
       3.33749622e+04, 3.14576506e+04, 2.92778276e+04, 2.86319154e+04,
       2.62354769e+04, 2.37310285e+04, 2.11589698e+04, 2.02927207e+04,
       1.92104980e+04, 1.74594467e+04, 1.50766308e+04, 1.45437234e+04,
       1.34818230e+04, 1.15812389e+04, 1.11343064e+04, 8.88943166e+03,
       8.52770701e+03, 7.84454791e+03, 7.01903957e+03, 6.28096534e+03,
       5.94189556e+03, 4.55129594e+03, 4.29449901e+03, 3.98799216e+03,
       3.48839564e+03, 3.07577637e+03, 1.99309231e+03, 1.53287832e+03,
       1.41769571e+03, 9.72051190e+02, 7.06002343e+02, 5.16084247e+02,
       4.81320083e+02, 2.75784281e+02, 1.36564925e+02, 8.98359317e+01,
       4.40150854e+01, 9.20662326e+00])

<IPython.core.display.Javascript object>

In [290]:
import time
import matplotlib.pyplot as plt


def plot_random_func(nd=50):
    fig = go.Figure()
    fig.add_trace(
        go.Scatter(
            x=list(range(nd)),
            y=np.random.multivariate_normal([0] * nd, get_nd_cov(nd)),
        )
    )
    fig.update_layout(title="Function as a random sample")
    fig.show()

<IPython.core.display.Javascript object>

## Random functions

In [306]:
plot_random_func()

<IPython.core.display.Javascript object>

# Functions as smooth random samples

In [650]:
import math


def get_nd_smooth_cov(d):
    npa = np.zeros(shape=(d, d))
    for i in range(d):
        for j in range(d):
            npa[i, j] = 4 * math.exp(-(abs(i - j) ** 2) / 75)
    return npa


<IPython.core.display.Javascript object>

In [651]:
# m1 = get_nd_smooth_cov(50)
# np.linalg.eig(m1)[0]

<IPython.core.display.Javascript object>

In [652]:
import time
import matplotlib.pyplot as plt


def plot_random_smooth_func(nd=50):
    fig = go.Figure()
    fig.add_trace(
        go.Scatter(
            x=list(range(nd)),
            y=np.random.multivariate_normal([0] * nd, get_nd_smooth_cov(nd)),
        )
    )
    fig.update_layout(title="Smooth function as a random sample")
    fig.show()

<IPython.core.display.Javascript object>

In [653]:
# np.linalg.eig(get_nd_smooth_cov(50))[0]

<IPython.core.display.Javascript object>

In [661]:
plot_random_smooth_func()

<IPython.core.display.Javascript object>

# (Empirical) Smooth functions as random samples but conditioned

In [370]:
def plot_random_smooth_func_conditioned(known_dict, nd=50, tol=0.1):
    fig = go.Figure()

    cov_matrix = get_nd_smooth_cov(nd)
    failed = True
    while failed:
        candidate = np.random.multivariate_normal([0] * nd, cov_matrix)
        failed = False
        for idx in known_dict.keys():
            if np.abs(candidate[idx] - known_dict[idx]) > tol:
                failed = True
                break

    known_points = list(known_dict.keys())
    known_values = [known_dict[x] for x in known_points]
    fig.add_trace(
        go.Scatter(
            x=list(range(nd)),
            y=candidate,
        )
    )
    fig.add_trace(go.Scatter(x=known_points, y=known_values, mode="markers"))
    fig.update_layout(
        title="Smooth function as a random sample conditioned on known points"
    )
    fig.show()

<IPython.core.display.Javascript object>

In [757]:
plot_random_smooth_func_conditioned({0: 0.5, 5: 1, 25: -1, 40: 1.5}, tol=0.5)

<IPython.core.display.Javascript object>

# (Theoretical) Smooth functions random samples but conditioned

In [910]:
def plot_random_smooth_func_conditioned_theoretical(known_dict, nd=50, n_output=1):
    fig = go.Figure()

    known_dimensions = list(known_dict.keys())
    unknown_dimensions = [x for x in range(nd) if x not in known_dimensions]
    known_values = [known_dict[dim] for dim in known_dimensions]

    assert len(known_dimensions) > 0

    cov_matrix = get_nd_smooth_cov(nd)

    # Need to find cov matrix for the conditional distribution
    # For this: we will apply the analytical formula
    #           So we need to first separate out the cov_matrix as A, B and C
    #           Refer RW.pdf conditional distribution (or see in the notebook above)
    #           to know what A, B and C are

    A = cov_matrix[unknown_dimensions, :][:, unknown_dimensions]
    B = cov_matrix[known_dimensions, :][:, known_dimensions]
    C = cov_matrix[unknown_dimensions, :][:, known_dimensions]

    #
    # Find conditional distribution parameters
    #
    new_mean = (
        np.array([0] * len(unknown_dimensions)).reshape(-1, 1)
        + C.dot(np.linalg.inv(B)).dot(np.array(known_values).reshape(-1, 1))
    )[:, 0]

    new_cov = A - C.dot(np.linalg.inv(B)).dot(np.transpose(C))

    for _ in range(n_output):
        unknown_values = np.random.multivariate_normal(new_mean, new_cov)
        unknown_dict = dict(zip(unknown_dimensions, unknown_values))

        merged_values = [
            known_dict[d] if d in known_dimensions else unknown_dict[d]
            for d in range(nd)
        ]

        fig.add_trace(go.Scatter(x=list(range(nd)), y=merged_values))
    fig.add_trace(
        go.Scatter(
            x=known_dimensions, y=known_values, mode="markers", marker={"color": "red"}
        )
    )
    fig.update_layout(
        title="Smooth function as a random sample conditioned on known points"
    )
    fig.show()

<IPython.core.display.Javascript object>

In [918]:
plot_random_smooth_func_conditioned_theoretical(
    {0: 0.5, 7: -1, 11: 1, 14: -2, 30: 1, 80: -10, 60: 3}, nd=100, n_output=1
)

<IPython.core.display.Javascript object>

# Gaussian process regression - A handcoded example

| Height | Weight | Age | BPsystolic | BPDiastolic | FastingSugarLevel | <span style="color:red">TimeToCompletePhysicalTask</span> |
| ---- | ---- | ---- |---- | ---- | ---- |---- |
| 6.23 | 66 | 34 | 120 | 89 | 342 | <span style="color:red">2.34 </span> |
| 5.2 | 56 | 37 | 145 | 80 | 242 | <span style="color:red">1.34</span> |
| 4.45 | 50 | 24 | 130 | 70 | 82 | <span style="color:red">2.54</span> |
| 7.10 | 89 | 40 | 101 | 65 | 202 | <span style="color:red">3.34</span> |
| 6.23 | 92 | 27 | 148 | 89 | 102 | <span style="color:red">4.34</span> |
| 6.95 | 123 | 47 | 153 | 86 | 294 | <span style="color:red">1.04</span> |
| 4.9 | 51 | 67 | 123 | 72 | 194 | <span style="color:red">2.54</span> |
| 5.2 | 56 | 25 | 115 | 74 | 142 | <span style="color:red">3.04</span> |
| 5.57 | 70 | 40 | 104 | 67 | 104 | <span style="color:red">2.74</span> |
| 5.9 | 62 | 89 | 109 | 90 | 243 | <span style="color:red">1.34</span> |
| 4.67 | 60 | 12 | 135 | 95 | 104 | <span style="color:red">1.94</span> |
| 6.59 | 70 | 41 | 145 | 69 | 120 |<span style="color:red"> 2.14</span> |
<br>

### Can we answer the following from the above??

<br>

| Height | Weight | Age | BPsystolic | BPDiastolic | FastingSugarLevel | <span style="color:red">TimeToCompletePhysicalTask</span> |
| ---- | ---- | ---- |---- | ---- | ---- |---- |
| 5.12 | 52 | 23 | 134 | 84 | 183 |<span style="color:red"> ? </span> |
| 6.2 | 78 | 43 | 143 | 93 | 135 |<span style="color:red"> ? </span> |

In [1395]:
known_dimensions = np.array(
    [
        [6.23, 66, 34, 120, 89, 342],
        [5.2, 56, 37, 145, 80, 242],
        [4.45, 50, 24, 130, 70, 82],
        [7.10, 89, 40, 101, 65, 202],
        [6.23, 92, 27, 148, 89, 102],
        [6.95, 123, 47, 153, 86, 294],
        [4.9, 51, 67, 123, 72, 194],
        [5.2, 56, 25, 115, 74, 142],
        [5.57, 70, 40, 104, 67, 104],
        [5.9, 62, 89, 109, 90, 243],
        [4.67, 60, 12, 135, 95, 104],
        [6.59, 70, 41, 145, 69, 120],
    ]
)

unknown_dimensions = np.array(
    [[5.12, 52, 23, 134, 84, 183], [6.2, 78, 43, 143, 93, 135]]
)

known_values = [2.34, 1.34, 2.54, 3.34, 4.34, 1.04, 2.54, 3.04, 2.74, 1.34, 1.94, 2.14]

<IPython.core.display.Javascript object>

In [1396]:
known_dimensions.shape

(12, 6)

<IPython.core.display.Javascript object>

In [1397]:
unknown_dimensions.shape

(2, 6)

<IPython.core.display.Javascript object>

In [1398]:
cov_base = np.cov(known_dimensions, unknown_dimensions)

<IPython.core.display.Javascript object>

In [1399]:
# unknown dims with unknown dims
A = cov_base[[12, 13], :][:, [12, 13]]
A.shape

(2, 2)

<IPython.core.display.Javascript object>

In [1400]:
# known with known
B = cov_base[0:12, :][:, 0:12]
B.shape

(12, 12)

<IPython.core.display.Javascript object>

In [1401]:
# unknown with known
C = cov_base[[12, 13], :][:, 0:12]
C.shape

(2, 12)

<IPython.core.display.Javascript object>

In [1402]:
# Find the conditional distribution
new_mean = (
    np.array([0] * 2).reshape(-1, 1)
    + C.dot(np.linalg.inv(B)).dot(np.array(known_values).reshape(-1, 1))
)[:, 0]

new_cov = A - C.dot(np.linalg.inv(B)).dot(np.transpose(C))


<IPython.core.display.Javascript object>

In [1403]:
new_mean

array([2.82468648, 2.81332251])

<IPython.core.display.Javascript object>

In [1404]:
new_cov

array([[ -929.98587014,  -798.86137568],
       [-2130.81348423, -1551.61161511]])

<IPython.core.display.Javascript object>

## Well... Thats kind of stupid!
#### But we can see why!

```
The Covariance numbers were inflated due to the "Glucose levels" which is on a much higher scale.
So the model believes there is so much variance and reflects the same in the result....
Either we should standardize (or) we should use a proper kernel
```

### Lets try RBF kernel with proper scale instead of np.cov
[RBF kernel formula](https://scikit-learn.org/stable/modules/generated/sklearn.gaussian_process.kernels.RBF.html)


In [1405]:
known_unknown_dimensions = np.concatenate(
    [known_dimensions, unknown_dimensions], axis=0
)

<IPython.core.display.Javascript object>

In [1406]:
known_unknown_dimensions.shape

(14, 6)

<IPython.core.display.Javascript object>

In [1407]:
def rbf_cov(npa, scale=100):
    cov_matrix = np.zeros(shape=(len(npa), len(npa)))
    for idx_i, i in enumerate(npa):
        for idx_j, j in enumerate(npa):
            dist = np.sum((i - j) ** 2)
            cov_matrix[idx_i, idx_j] = math.exp(-dist / (2 * scale))
    return cov_matrix

<IPython.core.display.Javascript object>

In [1416]:
rbf_cov_base = rbf_cov(known_unknown_dimensions, scale=1000)

<IPython.core.display.Javascript object>

In [1417]:
rbf_cov_base

array([[1.00000000e+00, 4.48045601e-03, 1.38915020e-15, 2.61573993e-05,
        1.46097820e-13, 3.30326372e-02, 7.81954296e-06, 1.66069367e-09,
        3.37160337e-13, 1.53066278e-03, 3.38681316e-13, 1.15185216e-11,
        2.47290731e-06, 3.37484865e-10],
       [4.48045601e-03, 1.00000000e+00, 2.11748552e-06, 8.79220194e-02,
        2.63634903e-05, 2.47722489e-02, 1.51291778e-01, 3.92652684e-03,
        2.62047871e-05, 1.26344244e-01, 4.51671525e-05, 4.95736486e-04,
        1.47341361e-01, 2.30770905e-03],
       [1.38915020e-15, 2.11748552e-06, 1.00000000e+00, 1.98442485e-04,
        2.39169159e-01, 6.26135187e-12, 7.29171755e-04, 1.43807274e-01,
        4.01267416e-01, 1.73593192e-07, 5.02065746e-01, 3.06729263e-01,
        5.46591113e-03, 9.75362427e-02],
       [2.61573993e-05, 8.79220194e-02, 1.98442485e-04, 1.00000000e+00,
        1.53093220e-03, 1.64994842e-03, 2.49718749e-01, 7.44740155e-02,
        6.80450457e-03, 6.38818497e-02, 1.29992014e-03, 1.08985007e-02,
        1.761

<IPython.core.display.Javascript object>

In [1418]:
# unknown dims with unknown dims
rbf_A = rbf_cov_base[[12, 13], :][:, [12, 13]]
rbf_A.shape

(2, 2)

<IPython.core.display.Javascript object>

In [1419]:
# known with known
rbf_B = rbf_cov_base[0:12, :][:, 0:12]
rbf_B.shape

(12, 12)

<IPython.core.display.Javascript object>

In [1420]:
# unknown with known
rbf_C = rbf_cov_base[[12, 13], :][:, 0:12]
rbf_C.shape

(2, 12)

<IPython.core.display.Javascript object>

In [1421]:
rbf_new_mean = (
    np.array([0] * 2).reshape(-1, 1)
    + rbf_C.dot(np.linalg.inv(rbf_B)).dot(np.array(known_values).reshape(-1, 1))
)[:, 0]

rbf_new_cov = rbf_A - rbf_C.dot(np.linalg.inv(rbf_B)).dot(np.transpose(rbf_C))


<IPython.core.display.Javascript object>

In [1422]:
rbf_new_mean

array([1.81874593, 2.09308664])

<IPython.core.display.Javascript object>

In [1423]:
rbf_new_cov

array([[0.77389829, 0.06074992],
       [0.06074992, 0.51671687]])

<IPython.core.display.Javascript object>

| Height | Weight | Age | BPsystolic | BPDiastolic | FastingSugarLevel | <span style="color:green">Predicted TimeToCompletePhysicalTask</span> |
| ---- | ---- | ---- |---- | ---- | ---- |---- |
| 5.12 | 52 | 23 | 134 | 84 | 183 |<span style="color:green"> 1.81 on avg with 0.77 variance </span> |
| 6.2 | 78 | 43 | 143 | 93 | 135 |<span style="color:green">  2.09 on avg with 0.51 variance </span> |